### PR TITLE
Resize large resolution PDFs to improve rendering time

### LINKED ISSUE #6575: Change code base to improve performance of PDFJS (see the issue https://github.com/mozilla/pdf.js/issues/6493)
All,

According to the issue https://github.com/mozilla/pdf.js/issues/6493, there is a problem that relating to performance when rendering a special scanning file (CCITTFaxStream). 

I used the latest version of pdf js (v1.1.469). I deployed to IIS Manager and tested it on localhost but pdf.js renders slowly.
Here is detail information:
- File for testing: https://drive.google.com/open?id=0BzoKdQwBinAnXzVJcmdVZjBmbkE
- System detail: my computer's hardware: Window 7x 64, RAM: 12 GB, Processors: Intel Core i5. I use Chrome browser and its version is 46.0.2490.80 m

I have logged time at render function of PDF.JS as follows:

``` javascript
render: function PDFPageProxy_render(params) {
      var stats = this.stats;
      stats.time('Overall');
      ....
    function complete(error) {
       ...
        stats.timeEnd('Rendering');
        stats.timeEnd('Overall');

        console.log(stats.toString());
      }
}
```

This is logging time:
Page Request: 20ms
Rendering: 3284ms
=> Overall: 3304ms

I would like to change the code base to improve performance. I have done the following things:
- I found that pdf.js takes long time at the function putBinaryImageData
- I have checked the function putBinaryImageData . This function renders an image into canvas from bytes array. Because bytes array is too big (16,843,200 bytes), this function takes about 3s to render. Thus, in order to reduce the processing time, I think the solution is to scale down the input bytes array before rendering. But it is unable to do with JavaScript.

Could you possibly advise how to improve in this case?

Regards,
Huy

### PR DESCRIPTION
**Pull request by hhstkh**
Fixes #6575.

### CODE REVIEW

File: src/core/image.js [hunk 1]
```diff
@@ -563,6 +666,12 @@ var PDFImage = (function PDFImageClosure() {
               buffer[i] ^= 0xff;
             }
           }
+
+          // Added by RamSoft (11/4/2015)
```
Comment (timvandermeij) [inline src/core/image.js]: Please remove these kinds of comments and set this as the author of the commit instead.

**Comment by timvandermeij** [comment]
After these changes, please squash the commits: https://github.com/mozilla/pdf.js/wiki/Squashing-Commits

File: src/core/image.js [hunk 2]
```diff
@@ -506,6 +506,109 @@ var PDFImage = (function PDFImageClosure() {
       }
     },
 
+    /**
+     * Added by RamSoft (11/5/2015) to resize large resolution PDFS 
+     * as to improve rendering time
+     */
+    resizeGrayPixels: function(imgData, comps, bpc) {
```
Comment (yurydelendik) [inline src/core/image.js]: create a separate resizeBWPixels function for bpc === 1 case

File: src/core/image.js [hunk 3]
```diff
@@ -506,6 +506,109 @@ var PDFImage = (function PDFImageClosure() {
...
+      var result = false;
+
+      //we only resize 1 bit per pixel or 8 bits per pixel grayscale image
+      if (comps === 1 && (bpc === 1 || bpc === 8)) {
```
Comment (yurydelendik) [inline src/core/image.js]: there is no need to check this here, move these checks out of the function and return new imgData instead

Comment (yurydelendik) [inline src/core/image.js]: e.g. shallResizeImage function might be added

File: src/core/image.js [hunk 4]
```diff
@@ -506,6 +506,109 @@ var PDFImage = (function PDFImageClosure() {
...
+        var scaleBits;
+        var h1 = imgData.height;
+        var w1 = imgData.width;
+
+        if ((h1 > 15000) || (w1 > 15000)) {
```
Comment (yurydelendik) [inline src/core/image.js]: document/comment how these magic numbers were selected

File: src/core/image.js [hunk 5]
```diff
@@ -506,6 +506,109 @@ var PDFImage = (function PDFImageClosure() {
...
+          var pixelArrayOutput = new Uint8Array(numBytes);
+          var x2, y2;
+          var originalRowStart;
+          var newRowStart;
+          var i, j;
+
+          if (bpc === 1) {
+
+            for (i = 0; i < h2; i++) {
+              y2 = i << scaleBits;
```
Comment (yurydelendik) [inline src/core/image.js]: here and below for i/x2 and j/y2 pairs `for (i = 0, y2 = 0; i < h2; i++, y2 += step) {` to remove `i << scaleBits` calculations (`var step = 1 << scaleBits`)

**Comment by hhstkh** [comment]
I have updated code as Yury's review. Please verify it.

File: src/core/image.js [hunk 6]
```diff
@@ -506,6 +506,122 @@ var PDFImage = (function PDFImageClosure() {
...
+     * @param  {Number} comps Number of color components, 1 or 3 is supported.
+     * @param  {Number} bpc Number of bits per component.
+     */
+    resizeGrayPixels: function PDFImage_resizeGrayPixels(imgData, comps, bpc) {
+      var scaleBits;
+      var h1 = imgData.height;
+      var w1 = imgData.width;
+
+      // Reduce the pixel dimensions to around 2000.
+      if ((h1 > 15000) || (w1 > 15000)) {
```
Comment (Snuffleupagus) [inline src/core/image.js]: https://github.com/mozilla/pdf.js/pull/6717#discussion_r46593022 still seems to be unresolved, since there are no comment explaining *how* these numbers were chosen.

File: src/core/image.js [hunk 7]
```diff
@@ -506,6 +506,122 @@ var PDFImage = (function PDFImageClosure() {
       }
     },
 
+    /**
+     * Check whether resize image data or not
+     * @param {Number} comps Number of color components, 1 or 3 is supported.
```
Comment (Snuffleupagus) [inline src/core/image.js]: This comment doesn't seem to be equivalent to the code, since there is no `comps === 3` case in the code.

File: src/core/image.js [hunk 8]
```diff
@@ -506,6 +506,122 @@ var PDFImage = (function PDFImageClosure() {
...
+     * @param {Number} bpc Number of bits per component.
+     * @return {Boolean}
+     */
+    shallResizeImage: function PDFImage_shallResizeImage(comps, bpc) {
+        return (comps === 1 && (bpc === 1 || bpc === 8));
```
Comment (Snuffleupagus) [inline src/core/image.js]: Nit: Here, and elsewhere in the patch, please make sure that you indent using *two* spaces.

File: src/core/image.js [hunk 9]
```diff
@@ -615,6 +763,7 @@ var PDFImage = (function PDFImageClosure() {
       return imgData;
     },
 
+
```
Comment (timvandermeij) [inline src/core/image.js]: Unnecessary change, please revert this line.

File: src/core/image.js [hunk 10]
```diff
@@ -506,6 +506,122 @@ var PDFImage = (function PDFImageClosure() {
...
+      var newRowBytes = (newWidth + 7) >> 3;
+      var originalRowBytes = (imgData.width + 7) >> 3;
+      var step = 1 << scaleBits;
+      var numBytes = newHeight * newRowBytes;
+      var pixelArrayOutput = new Uint8Array(numBytes);
+      var originalRowStart;
+      var newRowStart;
+
+      for (var i = 0, y2 = 0; i < newHeight; i++, y2 += step,
```
Comment (timvandermeij) [inline src/core/image.js]: This `for` loop condition is hard to understand. Please split this into multiple variables/lines.

File: src/core/image.js [hunk 11]
```diff
@@ -506,6 +506,122 @@ var PDFImage = (function PDFImageClosure() {
...
+     * Resize black white image data
+     * @param {Uint8Array} imgData image data.
+     * @param {Number} scaleBits
+     * @author Ramsoft
+     */
+    resizeBWPixels: function PDFImage_resizeBWPixels(imgData, scaleBits) {
+      var newWidth = imgData.width >> scaleBits;
+      var newHeight = imgData.height >> scaleBits;
+
+      var newRowBytes = (newWidth + 7) >> 3;
```
Comment (timvandermeij) [inline src/core/image.js]: `+ 7 >> 3` is a bit magical. Please add a comment to explain what this does exactly.

**Comment by timvandermeij** [comment]
After addressing these comments, please squash all the commits into one commit. See https://github.com/mozilla/pdf.js/wiki/Squashing-Commits on how to do that easily.

File: src/core/image.js (hunk 6 above)
Comment (hhstkh) [inline src/core/image.js]: These numbers based on experimentation. So I only add comments as above.

**Comment by hhstkh** [comment]
I merged into one commit. Please review my changes.

**Comment by timvandermeij** [comment]
There are still 30 commits here (see https://github.com/mozilla/pdf.js/pull/6717/commits). The steps mentioned at https://github.com/mozilla/pdf.js/wiki/Squashing-Commits will make that one commit.

**Comment by hhstkh** [comment]
Because I have some missing codes. Please continue to review at: https://github.com/mozilla/pdf.js/pull/7047.

Regards,

**Comment by jlukic** [comment]
:(
